### ISSUE TITLE
TypeError: Object(...) is not a function in depTrackingCache.js when Upgrading from 1.2.10  to 1.3.0

### ISSUE THREAD

**Issue by damiangreen** [apollographql/apollo-client#3990]
**Intended outcome:**
Upgrading from apollo-cache-inmemory ^1.2.10 to 1.3.0
Intend no issues in consuming code
**Actual outcome:**
Stack trace in browser
```TypeError: Object(...) is not a function
in /node_modules/apollo-cache-inmemory/lib/depTrackingCache.js:8
 this.depend = wrap(function (dataId) { return _this.data[dataId]; }, {

defaultNormalizedCacheFactory
return new DepTrackingCache(seed);
/node_modules/apollo-cache-inmemory/lib/depTrackingCache.js:60
```
**How to reproduce the issue:**

Consumer code  (unchanged) (originaly from React Starter Kit)
```export default function createCache() {
  // https://www.apollographql.com/docs/react/basics/caching.html#configuration
  return new InMemoryCache({
    dataIdFromObject,
  });
}
```

**Versions**

`System:
    OS: Windows 10
  Binaries:
    Yarn: 1.9.4 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 5.7.1 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: 42.17134.1.0
  npmPackages:
    apollo-cache-inmemory: 1.3.0 => 1.3.0
    apollo-client: 2.4.2 => 2.4.2
    apollo-link: ^1.0.7 => 1.2.1
    apollo-link-error: ^1.0.3 => 1.0.7
    apollo-link-http: ^1.3.2 => 1.5.3
    apollo-link-logger: ^1.1.0 => 1.2.0
    apollo-link-schema: ^1.0.1 => 1.0.6
    react-apollo: 2.1 => 2.1.9`

**Comment by benjamn** [comment]
Can you include more of the stack trace, including the error?

**Comment by damiangreen** [comment]
```
TypeError: Object(...) is not a function
new DepTrackingCache
D:/code/node_modules/apollo-cache-inmemory/lib/depTrackingCache.js:8
   5 | if (data === void 0) { data = Object.create(null); }
   6 | var _this = this;
   7 | this.data = data;
>  8 | this.depend = wrap(function (dataId) { return _this.data[dataId]; }, {
   9 |     disposable: true,
  10 |     makeCacheKey: function (dataId) {
  11 |         return dataId;
View compiled
defaultNormalizedCacheFactory
D:/code/node_modules/apollo-cache-inmemory/lib/depTrackingCache.js:60
  57 | }());
  58 | export { DepTrackingCache };
  59 | export function defaultNormalizedCacheFactory(seed) {
> 60 |     return new DepTrackingCache(seed);
  61 | }
  62 | //# sourceMappingURL=depTrackingCache.js.map
View compiled
new InMemoryCache
D:/code/node_modules/apollo-cache-inmemory/lib/inMemoryCache.js:69
  66 |     _this.config.cacheRedirects = _this.config.cacheResolvers;
  67 | }
  68 | _this.addTypename = _this.config.addTypename;
> 69 | _this.data = defaultNormalizedCacheFactory();
  70 | _this.storeReader = new StoreReader({
  71 |     addTypename: _this.config.addTypename,
  72 |     cacheKeyRoot: _this.cacheKeyRoot,
View compiled
createCache [as default]
D:/code/src/core/createApolloClient/createCache.ts:20
  17 | 
  18 | export default function createCache() {
  19 |   // https://www.apollographql.com/docs/react/basics/caching.html#configuration
> 20 |   return new InMemoryCache({
  21 |     dataIdFromObject,
  22 |   });
  23 | }
View compiled
./src/core/createApolloClient/createApolloClient.client.ts
D:/code/src/core/createApolloClient/createApolloClient.client.ts:40
  37 |   new HttpLink(resolveHttpLinkOptions()),
  38 | ]);
  39 | 
> 40 | const cache = createCache();
  41 | 
  42 | export default function createApolloClient() {
  43 |   return new ApolloClient({
View compiled
__webpack_require__
D:/code/webpack/bootstrap:763
  760 | };
  761 | 
  762 | // Execute the module function
> 763 | modules[moduleId].call(module.exports, module, module.exports, hotCreateRequire(moduleId));
  764 | 
  765 | // Flag the module as loaded
  766 | module.l = true;
View compiled
fn
D:/code/webpack/bootstrap:141
  138 | 		);
  139 | 		hotCurrentParents = [];
  140 | 	}
> 141 | 	return __webpack_require__(request);
  142 | };
  143 | var ObjectFactory = function ObjectFactory(name) {
  144 | 	return {
View compiled
./src/client.js
http://localhost:3001/assets/client.js:1239:112
__webpack_require__
D:/code/webpack/bootstrap:763
  760 | };
  761 | 
  762 | // Execute the module function
> 763 | modules[moduleId].call(module.exports, module, module.exports, hotCreateRequire(moduleId));
  764 | 
  765 | // Flag the module as loaded
  766 | module.l = true;
View compiled
fn
D:/code/webpack/bootstrap:141
  138 | 		);
  139 | 		hotCurrentParents = [];
  140 | 	}
> 141 | 	return __webpack_require__(request);
  142 | };
  143 | var ObjectFactory = function ObjectFactory(name) {
  144 | 	return {
View compiled
(anonymous function)
D:/code/src/clientLoader.js:1
> 1 | // polyfills
  2 | import main from './client';
  3 | 
  4 | function run() {
View compiled
./src/clientLoader.js
http://localhost:3001/assets/client.js:1545:30
__webpack_require__
D:/code/webpack/bootstrap:763
  760 | };
  761 | 
  762 | // Execute the module function
> 763 | modules[moduleId].call(module.exports, module, module.exports, hotCreateRequire(moduleId));
  764 | 
  765 | // Flag the module as loaded
  766 | module.l = true;
View compiled
fn
D:/code/webpack/bootstrap:141
  138 | 		);
  139 | 		hotCurrentParents = [];
  140 | 	}
> 141 | 	return __webpack_require__(request);
  142 | };
  143 | var ObjectFactory = function ObjectFactory(name) {
  144 | 	return {
View compiled
0
http://localhost:3001/assets/client.js:2970:18
__webpack_require__
D:/code/webpack/bootstrap:763
  760 | };
  761 | 
  762 | // Execute the module function
> 763 | modules[moduleId].call(module.exports, module, module.exports, hotCreateRequire(moduleId));
  764 | 
  765 | // Flag the module as loaded
  766 | module.l = true;
View compiled
checkDeferredModules
D:/code/webpack/bootstrap:43
  40 | 	}
  41 | 	if(fulfilled) {
  42 | 		deferredModules.splice(i--, 1);
> 43 | 		result = __webpack_require__(__webpack_require__.s = deferredModule[0]);
  44 | 	}
  45 | }
  46 | return result;
View compiled
(anonymous function)
D:/code/webpack/bootstrap:885
  882 | // add entry module to deferred list
  883 | deferredModules.push([0,"vendors"]);
  884 | // run deferred modules when ready
> 885 | return checkDeferredModules();
  886 | 
View compiled
(anonymous function)
```

**Comment by damiangreen** [comment]
So i'm debugging it and it looks like the `data ` param to `DepTrackingCache` is undefined  so `data = Object.create(null);` gets invoked. I'm not sure if this would cause the `wrap` function to barf, i can't step into the `wrap` function.. ho hum.

**Comment by MichalPrzybylski** [comment]
Hi, I've got the same issue.   
I'm using `"apollo-boost": "0.1.16",`  and I'm creating apollo client like this:
```typescript
const getClient (uri: string, headers: Headers)new ApolloClient({
    uri,
    headers
  });
```
 I'm building the application with webpack and I'm using `babel-loader` along with`@babel/preset-typescript` and my `.tsconfig` file is:
```json
{
  "compilerOptions": {
    "target": "esnext",
    "allowJs": true,
    "checkJs": false,
    "moduleResolution": "node",
    "jsx": "react",
    "skipLibCheck": true,
    "noEmit": true,
    "lib": ["dom", "esnext"]
  },
  "files": ["./src/index.tsx"],
  "exclude": ["node_modules"]
}
```
I'm getting exactly the same error

**Comment by MichalPrzybylski** [comment]
@damiangreen Thanks for walk around however I'd like to avoid deleting `src`, however I agree that this issue may be related to build process (module resolving). I'll try to provide some test-case repo on tomorrow

**Comment by benjamn** [comment]
Though this hardly needs to be said, Webpack gives you enough rope to tie yourself up in some pretty inscrutable knots. If you're recompiling the `src` directory instead of just using the `lib` code that we compiled, you're taking matters into your own hands, and your mileage may vary. Thanks for the thorough investigation, and I'm sorry about the time you had to sink into this!

**Comment by damiangreen** [comment]
I think @MichalPrzybylski's case may be slightly different. I've not had the capacity to do more digging but will update when i get a chance

**Comment by maxguzenski** [comment]
> I think @MichalPrzybylski's case may be slightly different. I've not had the capacity to do more digging but will update when i get a chance

I had this issue as well, I fixed it with: 
```javascript
// BEFORE
modules: [
    '.',
    '..',
    path.resolve(__dirname, 'src'),
    'node_modules'
],

// AFTER (remove dot and 2-dots)
modules: [
    path.resolve(__dirname, 'src'),
    'node_modules'
],
```

After that, remove all node_modules and re-install (yarn install)

**Comment by markfranco** [comment]
Seems like I have been having the same issue as well.

Looking into the generated file webpack is unable to require the wrap function from the optimism package. (It's been frustrating cause I work on a team of 4, and it only happens to my machine)

[
![screen shot 2018-11-29 at 2 00 36 pm](https://user-images.githubusercontent.com/12584082/49283293-7dcbda80-f4e5-11e8-8772-008ec7bd6161.png)
](url)

Does anyone else have the same thing occurring?

**Comment by damiangreen** [comment]
Yeah, I'm afraid i've not foudn time to debug it further, if you see my message above, mnaybe you can dig deeper.

**Comment by damiangreen** [comment]
@benjamn  does @markfranco's  suggestion  make sense? Is this something we could do on the main br anch?

**Comment by mnieber** [comment]
I got stuck with the same error at the same line:

    this.depend = Object(_optimism__WEBPACK_IMPORTED_MODULE_0__["wrap"])(function (dataId) { return _this.data[dataId]; }, {

I wonder though why the issue was closed. Are new users expected to apply the (rather obscure looking) work-around? I'm looking forward to switch to graphql later (it seems an impressive piece of work), but for the moment I will stick to my REST api.

**Comment by damiangreen** [comment]
I agree this issue really needs reopening

**Comment by microcipcip** [comment]
Stuck with this error as well, I have no idea how to fix it :(. I am using TS with next and react-apollo.

**Comment by damiangreen** [comment]
@markfranco  @benjamn  did you make headway with this?

**Comment by vpantiushenko** [comment]
@damiangreen , have you found the solution yet ?

**Comment by damiangreen** [comment]
No.. afraid not. I was hoping @benjamn  would fix this. I'm still waiting on upgrading until it gets fixed

**Comment by damiangreen** [comment]
Mysteriously this issue disappeared when we upgraded *al*l Apollo libraries (as opposed to upgrading this single library.

In case it helps others we upgrade to:

"apollo-cache-inmemory": "^1.5.1",
    "apollo-client": "2.5.1",
    "apollo-link": "^1.2.11",
    "apollo-link-batch-http": "^1.2.11",
    "apollo-link-error": "^1.1.10",
    "apollo-link-http": "^1.5.14",
    "apollo-link-logger": "^1.2.3",
    "apollo-link-schema": "^1.2.2",
